src/runtime/ranger/ranger_resource_policy.cpp

```diff
@@ -27,45 +28,178 @@ bool policy_item::match(const access_type &ac_type, const std::string &user_name
+                                     const std::string &default_database_name)
+{
+    // Check if it is denied by any policy in current resource.
+    auto check_res = do_check_ranger_resource_policy(policies,
+                                                     ac_type,
+                                                     user_name,
+                                                     md_type,
+                                                     database_name,
+                                                     default_database_name,
+                                                     policy_check_type::kDeny);
+    if (access_control_result::kDenied == check_res) {
+        return access_control_result::kDenied;
+    }
+
+    // Check if it is allowed by any policy in current resource.
+    check_res = do_check_ranger_resource_policy(policies,
+                                                ac_type,
+                                                user_name,
+                                                md_type,
+                                                database_name,
+                                                default_database_name,
+                                                policy_check_type::kAllow);
+    if (access_control_result::kAllowed == check_res) {
+        return access_control_result::kAllowed;
+    }
+
+    // The check that does not match any policy in current reosource returns false.
+    return access_control_result::kDenied;
+}
+
+access_control_result
+do_check_ranger_resource_policy(const std::vector<ranger_resource_policy> &policies,
+                                const access_type &ac_type,
+                                const std::string &user_name,
+                                const match_database_type &md_type,
+                                const std::string &database_name,
+                                const std::string &default_database_name,
+                                const policy_check_type &check_type)
+{
+    for (const auto &policy : policies) {
+        if (match_database_type::kNeed == md_type) {
+            // Lagacy table not match any database.
+            if (database_name.empty() && policy.database_names.count("*") == 0 &&
+                policy.database_names.count(default_database_name) == 0) {
+                continue;
             }
+            // New table not match any database.
+            if (!database_name.empty() && policy.database_names.count("*") == 0 &&
+                policy.database_names.count(database_name) == 0) {
+                continue;
+            }
+        }
+        auto check_status = policy.policies.policies_check(ac_type, user_name, check_type);
+        // When policy_check_type is 'kDeny' and in a 'deny_policies' and not in any
+        // 'deny_policies_exclude'.
+        if (policy_check_type::kDeny == check_type &&
+            policy_check_status::kDenied == check_status) {
```

Comment:
If check_status equals to kDenied, check_type must be kDeny, right? If it is, refact the code like:
```
if (policy_check_status::kDenied == check_status) {
  CHECK_TRUE(policy_check_type::kDeny == check_type);
  return access_control_result::kDenied;
}
```